examples/migration/recursive_fibonacci/task_emulation_layer.h

```diff
@@ -0,0 +1,145 @@
+/*
+    Copyright (c) 2023 Intel Corporation
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+#ifndef __TBB_task_emulation_layer_H
+#define __TBB_task_emulation_layer_H
+
+#include "tbb/task_group.h"
+#include "tbb/task_arena.h"
+
+#include <atomic>
+
+namespace task_emulation {
+
+struct task_group_pool {
+    task_group_pool() : pool_size(std::thread::hardware_concurrency()), task_submiters(new tbb::task_group[pool_size]) {}
+
+    ~task_group_pool() {
+        for (std::size_t i = 0; i < pool_size; ++i) {
+            task_submiters[i].wait();
+        }
+
+        delete [] task_submiters;
+    }
+
+    tbb::task_group& operator[] (std::size_t idx) { return task_submiters[idx]; }
+
+    std::size_t pool_size;
+    tbb::task_group* task_submiters;
+};
+
+static task_group_pool tg_pool;
+
+class base_task {
+public:
+    base_task() = default;
+
+    base_task(const base_task& t) : m_parent(t.m_parent), m_child_counter(t.m_child_counter.load())
+    {}
+
+    virtual ~base_task() = default;
+
+    void operator() () const {
+        base_task* parent_snapshot = m_parent;
+        const_cast<base_task*>(this)->execute();
+        if (m_parent && parent_snapshot == m_parent) {
+            m_parent->release();
+        }
+    }
+
+    virtual void execute() = 0;
+
+    template <typename C, typename... Args>
+    C* create_continuation(std::uint64_t ref, Args&&... args) {
+        C* continuation = new C{std::forward<Args>(args)...};
+        continuation->set_parent(release_parent());
+        continuation->m_child_counter = ref;
+        return continuation;
+    }
+
+    template <typename F, typename... Args>
+    F create_child_of_continuation(Args&&... args) {
+        F obj{std::forward<Args>(args)...};
+        obj.set_parent(this);
+        return obj;
+    }
+
+    template <typename C>
+    void recycle_as_child_of_continuation(C& c) {
+        set_parent(&c);
+    }
+
+protected:
+    void reserve() {
+        ++m_child_counter;
+    }
+
+private:
+    template <typename F, typename... Args>
+    friend F create_root_task(tbb::task_group& tg, Args&&... args);
+
+
+    void release() {
+        if (--m_child_counter == 0) {
+            operator()();
```

Comment:
I'm not sure if such overflow is possible in reasonable scenarios because in tree-like problems the deps of stack recursion will be no more then deps of tree.
However, recursion cold be avoided with new task spawn. Do you think is it make sense to add such comment in emulation internals?

Reply:
Perhaps. May be even implement it?

Reply:
I'm not sure because it will reduce performance dramatically.

Reply:
For the migration example, why are there the performance measurements at all? I believe their  purpose is to show how to do the migration.
Do all the migration examples have performance measurements?

Reply:
The idea is to provide efficient emulation layer that would be a replacement for part of the Task API.